android/src/main/java/com/swmansion/reanimated/nodes/CallFuncNode.java

```diff
@@ -0,0 +1,46 @@
+package com.swmansion.reanimated.nodes;
+
+import com.facebook.react.bridge.ReadableMap;
+import com.swmansion.reanimated.NodesManager;
+import com.swmansion.reanimated.Utils;
+
+public class CallFuncNode extends Node {
+
+    private int mWhatNodeID;
+    private int[] mArgs;
+    private int[] mParams;
+
+    public CallFuncNode(int nodeID, ReadableMap config, NodesManager nodesManager) {
+        super(nodeID, config, nodesManager);
+        mWhatNodeID = config.getInt("what");
+        mParams = Utils.processIntArray(config.getArray("params"));
+        mArgs = Utils.processIntArray(config.getArray("args"));
+    }
+
+    private void beginContext() {
+        mNodesManager.updateContext.contextCount++;
+        for(int i=0; i<mParams.length; i++){
+            int paramId = mParams[i];
+            ParamNode paramNode = mNodesManager.findNodeById(paramId, ParamNode.class);
+            paramNode.beginContext(mArgs[i]);
+        }
+    }
+
+    private void endContext() {
+        for(int i=0; i<mParams.length; i++){
+            int paramId = mParams[i];
+            ParamNode paramNode = mNodesManager.findNodeById(paramId, ParamNode.class);
+            paramNode.endContext();
+        }
+        mNodesManager.updateContext.contextCount--;
+    }
+
+    @Override
+    protected Object evaluate() {
+        beginContext();
+        Node whatNode = mNodesManager.findNodeById(mWhatNodeID, Node.class);
+        Object retVal = whatNode.value();
```

Comment:
maybe we can pass context here? And then taker advantages of memoization per context?

Reply:
I really didn't want an external evaluation context.... To try to avoid changing too much of the code - what is the evaluation context anyways? :) What I'm using is the param node that wraps another node - and for each function call it's inner node is switched to the id of the current parameter. This way things will be correct. We only need to ensure that calling the same function twice (recursive), the function should be evaluated each time.